### ISSUE TITLE
Add wheel distribution(s) of Astropy

### ISSUE THREAD

**Issue by embray** [astropy/astropy#802]
PEP-427 defining the Wheel format has finally been accepted after months of deliberation. This is finally the first ever documented and blessed binary distribution format for Python (it's a little like the .egg format but should not be confused with it--it's for distribution _only_ and is not designed so that packages can be directly imported from .whl files).

A reference implementation is here: https://bitbucket.org/dholth/wheel

Plans are afoot to add support to pip now that the PEP has been accepted.

Eventually there will be some stdlib support for wheel as well, possibly as part of the ongoing distilb project.

So it would be worth embracing the future at some point providing Wheel distributions for all the platforms we support (including Linux).

**Comment by astrofrog** [comment]
I need to look this up, but is it possible to build wheels for different versions of numpy? (i.e. one for 1.6, 1.7, etc.).

**Comment by astrofrog** [comment]
We now usually have wheels for MacOS X thanks to @matthew-brett:

https://github.com/MacPython/astropy-wheels

I wonder whether it would make sense, since Travis now allows multi-os builds, to extend this to become a build repository for wheels that does Linux, MacOS X, and Windows wheels?

@matthew-brett @mwcraig - what do you think?

**Comment by matthew-brett** [comment]
How timely!  I just extended that repo to build and test manylinux wheels : https://travis-ci.org/MacPython/astropy-wheels/builds/137716186

For windows I guess the easiest would be to add an appveyor.yml file to that repo.  Is there an astropy appyveyor account?

**Comment by astrofrog** [comment]
@matthew-brett - yes, an appveyor.yml file would be good. There is an AppVeyor astropy team - the main administrator is astropy.team@gmail.com. I can add you to the team so you can add the repo there if you like? (if so, what email should I use?)

**Comment by astrofrog** [comment]
@matthew-brett - thanks for working on this by the way!

As a side note, I've been also working on making some wheels for ARM/Raspberry pi, but I think those will have to be done manually :)

**Comment by matthew-brett** [comment]
My email is matthew.brett@gmail.com - please add me, and I'll make up an appveyor file.

I have a builtbot slave running on a raspberry pi - maybe you could do something like that?

Did you see https://hub.docker.com/r/sdt4docker/raspberry-pi-cross-compiler/ ?

**Comment by mwcraig** [comment]
I'm fine with it, especially since @matthew-brett is willing to take the lead 😀 . I know the linux ones in particular are a little fussy to get right, so if he has a builder machine ready to go that would be awesome.

**Comment by astrofrog** [comment]
@matthew-brett - I've added you to the team - you'll need to log out and log back in again and select 'Astropy' when logging in.

Cool about the buildbot - do you just get it to look for the latest version of PyPI and do a build if there is a new version? Or do you manually trigger it?

By the way, does ``python setup.py build_wheel upload`` work?

**Comment by matthew-brett** [comment]
For the buildbot - its set up to trigger for each commit to the repos it's watching, and it can upload build products to a web server we have in a server room here in Berkeley.   If you'd like to use it, let me know - it would involve working out some buildbottiness at the build server setup [here](https://github.com/nipy/nibotmi/blob/master/master.cfg) but I would be happy to help.

Not sure what you mean about `build_wheel upload`?   The wheels from the MacPython get uploaded to http://wheels.scipy.org by the travis yml script.

**Comment by astrofrog** [comment]
@matthew-brett - sorry, there was a typo in the command - what I meant is that normally one can do:

```
python setup.py sdist upload
```

so I was just curious as to whether

```
python bdist_wheel upload
```

would work (but this is off-topic)

**Comment by matthew-brett** [comment]
Yes, that works in general - but when I'm working on a local machine, I personally prefer to build anyhow, maybe `python setup.py bdist_wheel` and then upload with `twine`.

**Comment by srkunze** [comment]
+1 for this issue

**Comment by tomkooij** [comment]
What's the status on this issue? windows wheel on PyPI would be very "nice to have".

As astropy already has appveyor CI running, it is quite easy to create windows wheels with appveyor: [Example build.](https://ci.appveyor.com/project/tomkooij/astropy/build/job/lwwa5j6x3pdbjta1) [Code.](https://github.com/tomkooij/astropy/commit/3ba11d1f4e9b8b0e66b6ae3b3ed4795b76184e64) Note that this only builds a wheel and skips installing from the wheel and running test.

I'll be happy to help out!

**Comment by mwcraig** [comment]
Fine question...looks like they are getting built and uploaded for mac and linux by @matthew-brett. I'm guessing they are done manually rather than automatically.

**Comment by matthew-brett** [comment]
No - they are done automatically - see :
https://github.com/MacPython/astropy-wheels

It would just be a question of adding an Appveyor setup there.

Tom K - should I add you to the MacPython repo to let you get going with
that?
​

**Comment by tomkooij** [comment]
@matthew-brett I've been doing some work on a fork of https://github.com/MacPython/pytables-wheels trying to get the multibuild bash scripts running in cygwin op appveyor.  That works, although it's a bit awkward using bash on windows and installing dependencies using `pip` is a pain on windows. 

It's probably much easier to use powershells scripts and  `conda` on appveyor. See for example: https://github.com/tomkooij/pytables-wheel-builder
It shouldn't be too difficult to cook up powershell scripts to match the multibuild workflow. That would then be easier to maintain and use for other repos.

Anyway, feel free to add me and I'll see if I can make something work.

**Comment by matthew-brett** [comment]
Great - if you could cook up something, that would be a huge help.   I've added you.

**Comment by tomkooij** [comment]
Linux, OSX and windows wheels are built for astropy-1.3 and are on https://wheels.scipy.org/. 

@matthew-brett will you upload these to pypi? (This is "nice to have" for my particular use case)

**Comment by matthew-brett** [comment]
Done - thanks.

**Comment by astrofrog** [comment]
Now that this is working (thanks @matthew-brett and @tomkooij!) I think the only step before we can close this would be to add a mention of this in the release documentation for astropy. I think just a couple of sentences would do. Anyone volunteer to add this? :)

**Comment by tomkooij** [comment]
I've started the builds for v1.3.1 on my local fork, if the builds pass, I'll push a PR to the MacPython repo, so we can have v1.3.1 builds soon. 

I'll have a look at the release docs as well 👍

**Comment by bsipocz** [comment]
Thanks @tomkooij.

**Comment by bsipocz** [comment]
The wheel building is now part of the release instructions, so I'm closing this off. Please reopen if you find it's not fulfil the aims here.
